src/modules/site-v2/base/views/data/data.py

```diff
@@ -57,3 +76,186 @@ def protocols():
     }
     return render_template('data/protocols.html', **params)
 
+
+#
+# Submit Trait
+#
+@data_bp.route('/submit-trait')
+@jwt_required()
+def submit_trait_start():
+  """ Submit Trait start page """
+  return render_template('data/submit-trait-start.html', **{
+    'title': 'Submit Trait',
+    'disable_parent_breadcrumb': True
+  })
+
+#
+# Submit Trait Form
+#
+@data_bp.route('/submit-trait/new-submission', methods=['GET', 'POST'])
+@jwt_required()
+def submit_trait_form():
+  """ Trait Submission Form """
+  form = TraitSubmissionForm()
+  user = get_current_user()
+
+  if hasattr(user, 'username') and not form.username.data:
+    form.username.data = user.username
+
+  # Handle form submission
+  if request.method == 'POST':
+    form = TraitSubmissionForm(request.form)
+    form.file.data = request.files.get('file')
+
+    # Validate form fields
+    if not form.validate_on_submit():
+      flash('Please fill out all required fields.', 'warning')
+    
+    else:
+      try:
+        # Create a new TraitFile object
+        tf = TraitFile(unique_id())
+
+        # Create a unique filename for file upload
+        hashed_filename = f'{unique_id()}.tsv'
+
+        tf.set_properties(**{
+          # User submitted data
+          'trait_name_user':      bleach.clean(form.trait_name_user.data),
+          'trait_name_display_1': bleach.clean(form.trait_name_display_1.data),
+          'trait_name_display_2': bleach.clean(form.trait_name_display_2.data),
+          'trait_name_display_3': bleach.clean(form.trait_name_display_2.data),
+          'filename':             bleach.clean(form.file.data.filename),
+          'species':              bleach.clean(form.species.data),
+          'description_short':    bleach.clean(form.description_short.data),
+          'description_long':     bleach.clean(form.description_long.data),
+          'units':                bleach.clean(form.units.data),
+          'tags':                 [ bleach.clean(tag) for tag in form.tags.data ],
+          'institution':          bleach.clean(form.institution.data),
+          'source_lab':           bleach.clean(form.source_lab.data),
+          'protocols':            bleach.clean(form.protocols.data),
+          'publication':          bleach.clean(form.publication.data),
+  
+          # Internally used data
+          'dataset':           'public',
+          'publish_status':    PublishStatus.UPLOADED,
+          'is_bulk_file':      False,
+          'hashed_filename':   hashed_filename,
+        })
+
+        tf.set_user(user)
+
+        # Save the TraitFile object to Datastore
+        tf.save()
+
+        # Seed to Phenotype Metadata SQL table
+        new_trait = PhenotypeMetadata()
+        new_trait.add_trait(tf)
+
+      except Exception as ex:
```

Comment:
I like that we are handling exceptions at all levels, but I wonder if we need to 'rollback' in case the DataStore succeeds, but the SQL operation fails -- should we remove the DataStore entry in that case? If we keep the entry and the user retries the operation, will that lead to 'ghost' traits in the datastore?

Reply:
That's a good point. I'll implement this.

Reply:
this is a good candidate for the `command pattern`: https://www.patterns.dev/vanilla/command-pattern/

Reply:
Pushed in the commit b96595d18c0d3fd589ea8793398ff122082faad0